geti_sdk/deployment/inference_hook_interfaces.py

```diff
@@ -0,0 +1,263 @@
+            else:
+                constructor_args.update({key: value})
+        return {type(self).__name__: constructor_args}
+
+    @classmethod
+    @abstractmethod
+    def from_dict(cls, input_dict: Dict[str, Any]) -> "PostInferenceObject":
+        """
+        Construct a PostInferenceObject from an input dictionary `input_dict`
+
+        :param input_dict: Dictionary representation of the PostInferenceObject
+        :return: Instantiated PostInferenceObject, according to the input dictionary
+        """
+        return NotImplemented
+
+
+class PostInferenceTrigger(PostInferenceObject):
+    """
+    Base class for post inference triggers
+
+    Post inference triggers are used in inference hooks. They define a condition
+    (or set of conditions) that determines whether any post-inference action should be
+    taken.
+
+    :param threshold: Threshold to use in activating the trigger. If the computed
+        score is above the threshold, the trigger emits a positive decision and any
+        downstream actions will be performed
+    """
+
+    def __init__(self, threshold: float = 0.5):
+        super().__init__()
+        self.threshold = threshold
+        self._repr_info_ = f"threshold={threshold:.1f}"
+
+    @abstractmethod
+    def __call__(self, image: np.ndarray, prediction: Prediction) -> float:
+        """
+        Compute a trigger score for the `image` and corresponding `prediction`.
+
+        :param image: Numpy array representing an image
+        :param prediction: Prediction object corresponding to the inference result
+            for the image.
+        :return: Float representing the score for the input
+        """
+        return NotImplemented
+
+    def get_decision(self, score: float) -> bool:
+        """
+        Make a decision based on a previously computed `score` and the threshold defined
+        for the trigger
+
+        :param score: Float representing the score for a certain image, prediction pair
+        :return: Boolean indicating whether the trigger conditions are met (True,
+            score is higher than the threshold), or not (False)
+        """
+        return score > self.threshold
+
+    def __repr__(self) -> str:
+        """
+        Return string representation of the PostInferenceTrigger
+        """
+        return f"{type(self).__name__}({self._repr_info_})"
+
+    @classmethod
+    def from_dict(cls, input_dict: Dict[str, Any]) -> "PostInferenceTrigger":
+        """
+        Construct a PostInferenceTrigger from an input dictionary `input_dict`
+
+        :param input_dict: Dictionary representation of the PostInferenceTrigger
+        :return: Instantiated PostInferenceTrigger, according to the input dictionary
+        """
+        available_triggers = {
+            subcls.__name__: subcls for subcls in cls.__subclasses__()
+        }
+        triggers: List["PostInferenceTrigger"] = []
+        for trigger_name, trigger_args in input_dict.items():
```

Comment:
In principle yes, unless a trigger takes non-serializable objects as input arguments. In that case it needs to implement a `from_dict` method that initializes the required objects. 

As I'm writing this I notice that this is overly complicated, I think it would be possible to construct triggers with serializable arguments only. That would save a lot of trouble in the saving and loading of the deployment.